sakura_core/dlg/CDlgJump.cpp

```diff
@@ -257,12 +257,12 @@ void CDlgJump::SetData( void )
 			if( m_pShareData->m_bLineNumIsCRLF_ForJump ){	/* 行番号の表示 false=折り返し単位／true=改行単位 */
 				auto_sprintf( szText, LS(STR_DLGJUMP_PSLQL),
 					cFuncInfoArr.GetAt( i )->m_nFuncLineCRLF,
-					cFuncInfoArr.GetAt( i )->m_cmemFuncName.GetStringPtr()
+					cFuncInfoArr.GetAt( i )->m_cmemFuncName.c_str()
```

Comment:
なんか、レビューする意欲がわかないのですが、こう書けるようにしたら今よりマシになるかと思います。
```suggestion
					cFuncInfoArr.GetAt( i ).GetFuncName()
```

- GetAt(n) の戻り値をポインタから参照に変えれば、nullptr考慮が不要になります。
- LPCWSTR を返すうんこgetterを作れば、外部に影響を与えずにメンバ型を変更できるようになります。

Reply:
もし `CFuncInfoArr::GetAt` の戻り値の型をポインタから参照に変えるとなると、引数の `nIdx` が範囲外だった場合に例外を送る必要が出てくるかもしれないと思いました。下記のような感じでしょうか？

```cpp
/* 0<=の指定番号のデータを返す */
/* データがない場合は例外を送出する */
CFuncInfo& CFuncInfoArr::GetAt( int nIdx )
{
	if( nIdx < 0 || nIdx >= m_cFuncInfoArr.size() ){
		throw std::out_of_range(__func__);
	}
	return m_cFuncInfoArr[nIdx];
}
```

なお `CFuncInfoArr::GetAt` の戻り値が nullptr じゃないかの確認は `CDlgFuncList::CompareFunc_Asc` では行っていますが他では行っていないですね。それで問題が無いのかというとちょっと心配な箇所もありますがおそらく大丈夫なんでしょう。

さて戻り値をポインタから参照に変えれば、nullptr の考慮は不要にはなると思いますが try catch を使った例外処理の記述が不十分だと回復できなくてアプリが落ちてしまいます。ただそれは現状の実装でも nullptr だった場合には落ちるのでそういう意味ではいっしょかな？

ただ `CFuncInfoArr::GetAt` の呼び出しの外で try catch を書く手間を考えたら、`CFuncInfoArr::GetNum` を呼んで事前に問題が無いindexなのかどうか調べる方が良い気がしてきました。

Reply:
実装するならnullオブジェクトだと思います。
例外にすると、説明してもらったような問題がでるので。

Reply:
```cpp
LPCWSTR CFuncInfo::GetFuncName() const;
```
を追加する方が確かにコードの見栄えが良くなりますね。

Reply:
> あと、このPRの趣旨は「メモリ確保を削減する」だと思うんですが、
> cNative.AppendNativeData(wstring.cstr())はマズいと思います。
> 
> 暗黙的に cNative.AppendNativeData(CNativeW(wstring.cstr())) になるからです。

これはその通りですね。ご指摘ありがとうございます。何も考えずに元のコードのまま `AppendNativeData` を使うようにしていました。後で `AppendString` を使うように修正を行います。

Reply:
> リファクタリングが成立するのか分からんです。
> 
> テストを用意しないと、変更前後でふるまいが変わらないことを証明できないです。
> 
> 「テスト不可コードなので変更前のテストは書けません」はしょうがない気がするんですけど、そういう変更はたぶんリファクタリングじゃないような気がします。

リファクタリングする際はテストを書いて変更前と処理結果が一致する事を確認していきながら進めた方が良いよってマーティン・ファウラーの本でも書かれてますね。手抜きしてテストは極力書いていませんが、ちゃんとテストしないと不具合がぽこぽこ出てしまうので反省してなるべく追加したいと思います。

なおリファクタリングの定義については個々人の心の中にあると思うので、別にテストを書かなくてもリファクタリングだとは思ってますが、他の人がそれを受け入れてくれるかは別ですね…。

Reply:
`GetAt` の戻り値を参照に変える変更を行いましたがそれに付随してかなり差分は増えました。
目視でのコードレビューはかなり大変そうです。

なお単体テストを用意しても変更前後で動作が変わらない事の証明にはなりません。一定の条件での確認にはなると思います。

単体テストが用意されてないとテストが自動再現出来ないしテストのエビデンスも残りにくいので、テスト追加しないと変更しちゃ駄目だよという意見はまぁもっともなんですけど、なんか気軽に変更しづらいなぁと思います。

Reply:
「リファクタリング」については、設計のリファクタリングをどう扱うかに迷いがあるためにこんな話し方になっています。単体テストを用意して「変更前後で挙動に差はありません」を担保するスタイルだと、「インターフェースが変」なのを対策するリファクタリングができないです。サクラエディタで「やったほうが良さそう」なネタとしてはインターフェース変更を伴うリファクタリングが多くて、しかもそういう変更は変更前がテスト不能である場合が多い気がします。


> なお単体テストを用意しても変更前後で動作が変わらない事の証明にはなりません。一定の条件での確認にはなると思います。

その通りだと思います。

「できるだけ単体テストを用意しようぜ！」と言ってる目的の一つは、
「テストが書けない」って事実に気付けるようにすることです。

検証不能なロジックは、検証コードを書いても結局検証できないっす。

書けないなりにテストを書いてみて、そのテストじゃ動作検証が十分に行えないと分かったならそれは対策すべきだと思うんですが、たぶん絶対まだそんな段階には達していないと思うので検証が行えてないのは一旦スルーでよいような気がします。ノリ的に「とりあえず努力はそとこうぜ」です。

Reply:
このPRでは色々なファイルを変更していますが、`CDialog` を継承した `CDlgFuncList` と `CDlgJump` についてはGUIなので単体テストで試験しにくそうなので手動で動作確認を念入りに行うようにします。

アウトライン解析 データ配列のクラスである `CFuncInfoArr` についてはまだテストコードが無いようですが、単体テストが行いやすい内容だと思うので後で追加を行います。